File: tests/python/test_vpwma.py
"""
Python binding tests for VPWMA indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestVpwma:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_vpwma_partial_params(self, test_data):
        """Test VPWMA with partial parameters (None values) - mirrors check_vpwma_partial_params"""
        close = test_data['close']
        
        
        result = ta_indicators.vpwma(close, 14, 0.382)  
        assert len(result) == len(close)
    
    def test_vpwma_accuracy(self, test_data):
        """Test VPWMA matches expected values from Rust tests - mirrors check_vpwma_accuracy"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['vpwma']
        
        result = ta_indicators.vpwma(
            close,
            period=expected['default_params']['period'],
            power=expected['default_params']['power']
        )
        
        assert len(result) == len(close)
        
        
        assert_close(
            result[-5:], 
            expected['last_5_values'],
            rtol=1e-4,  
            msg="VPWMA last 5 values mismatch"
        )
        
        
        compare_with_rust('vpwma', result, 'close', expected['default_params'])
    
    def test_vpwma_zero_period(self):
        """Test VPWMA fails with zero period - mirrors check_vpwma_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.vpwma(input_data, period=0, power=0.382)
    
    def test_vpwma_period_exceeds_length(self):
        """Test VPWMA fails when period exceeds data length - mirrors check_vpwma_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.vpwma(data_small, period=10, power=0.382)
    
    def test_vpwma_very_small_dataset(self):
        """Test VPWMA fails with insufficient data - mirrors check_vpwma_very_small_dataset"""
        single_point = np.array([42.0])
        
        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.vpwma(single_point, period=2, power=0.382)
    
    def test_vpwma_empty_input(self):
        """Test VPWMA fails with empty input"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="Input data slice is empty"):
            ta_indicators.vpwma(empty, period=14, power=0.382)
    
    def test_vpwma_invalid_power(self):
        """Test VPWMA fails with invalid power - mirrors vpwma power validation"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
        
        
        with pytest.raises(ValueError, match="Invalid power"):
            ta_indicators.vpwma(data, period=2, power=float('nan'))
        
        
        with pytest.raises(ValueError, match="Invalid power"):
            ta_indicators.vpwma(data, period=2, power=float('inf'))
    
    def test_vpwma_reinput(self, test_data):
        """Test VPWMA applied twice (re-input) - mirrors check_vpwma_reinput"""
        close = test_data['close']
        
        
        first_result = ta_indicators.vpwma(close, period=14, power=0.382)
        assert len(first_result) == len(close)
        
        
        second_result = ta_indicators.vpwma(first_result, period=5, power=0.5)
        assert len(second_result) == len(first_result)
        
        
        if len(second_result) > 240:
            for i in range(240, len(second_result)):
                assert not np.isnan(second_result[i])
    
    def test_vpwma_nan_handling(self, test_data):
        """Test VPWMA handles NaN values correctly - mirrors check_vpwma_nan_handling"""
        close = test_data['close']
        
        result = ta_indicators.vpwma(close, period=14, power=0.382)
        assert len(result) == len(close)
        
        
        if len(result) > 50:
            assert not np.any(np.isnan(result[50:])), "Found unexpected NaN after warmup period"
        
        
        assert np.all(np.isnan(result[:13])), "Expected NaN in warmup period"
    
    def test_vpwma_warmup_period(self, test_data):
        """Test VPWMA warmup period behavior - matches real-world conditions"""
        close = test_data['close']
        
        
        test_cases = [
            {'period': 5, 'power': 0.5, 'expected_warmup': 4},
            {'period': 10, 'power': 0.382, 'expected_warmup': 9},
            {'period': 14, 'power': 0.382, 'expected_warmup': 13},
            {'period': 20, 'power': 0.3, 'expected_warmup': 19}
        ]
        
        for tc in test_cases:
            result = ta_indicators.vpwma(close, period=tc['period'], power=tc['power'])
            
            
            warmup_values = result[:tc['expected_warmup']]
            assert np.all(np.isnan(warmup_values)), \
                f"Expected all NaN in warmup period for period={tc['period']}, got non-NaN values"
            
            
            if len(result) > tc['expected_warmup']:
                assert not np.isnan(result[tc['expected_warmup']]), \
                    f"Expected first valid value at index {tc['expected_warmup']} for period={tc['period']}"
            
            
            
            assert tc['expected_warmup'] == tc['period'] - 1, \
                f"Warmup calculation mismatch for period={tc['period']}"
    
    def test_vpwma_streaming(self, test_data):
        """Test VPWMA streaming matches batch calculation - mirrors check_vpwma_streaming"""
        close = test_data['close']
        period = 14
        power = 0.382
        
        
        batch_result = ta_indicators.vpwma(close, period=period, power=power)
        
        
        stream = ta_indicators.VpwmaStream(period=period, power=power)
        stream_values = []
        
        for price in close:
            result = stream.update(price)
            stream_values.append(result if result is not None else np.nan)
        
        stream_values = np.array(stream_values)
        
        
        assert len(batch_result) == len(stream_values)
        
        
        for i, (b, s) in enumerate(zip(batch_result, stream_values)):
            if np.isnan(b) and np.isnan(s):
                continue
            assert_close(b, s, rtol=1e-9, atol=1e-9, 
                        msg=f"VPWMA streaming mismatch at index {i}")
    
    def test_vpwma_batch(self, test_data):
        """Test VPWMA batch processing - mirrors check_batch_default_row"""
        close = test_data['close']
        
        result = ta_indicators.vpwma_batch(
            close,
            period_range=(14, 14, 0),  
            power_range=(0.382, 0.382, 0.0)  
        )
        
        assert 'values' in result
        assert 'periods' in result
        assert 'powers' in result
        
        
        assert result['values'].shape[0] == 1
        assert result['values'].shape[1] == len(close)
        
        
        default_row = result['values'][0]
        expected = EXPECTED_OUTPUTS['vpwma']['last_5_values']
        
        
        assert_close(
            default_row[-5:],
            expected,
            rtol=1e-4,
            msg="VPWMA batch default row mismatch"
        )
    
    def test_vpwma_batch_multiple_params(self, test_data):
        """Test VPWMA batch with multiple parameter combinations"""
        close = test_data['close'][:100]  
        
        
        result = ta_indicators.vpwma_batch(
            close,
            period_range=(10, 14, 2),    
            power_range=(0.3, 0.4, 0.1)  
        )
        
        assert 'values' in result
        assert 'periods' in result
        assert 'powers' in result
        
        
        assert result['values'].shape[0] == 6
        assert result['values'].shape[1] == 100
        assert len(result['periods']) == 6
        assert len(result['powers']) == 6
        
        
        expected_periods = [10, 10, 12, 12, 14, 14]
        expected_powers = [0.3, 0.4, 0.3, 0.4, 0.3, 0.4]
        
        np.testing.assert_array_equal(result['periods'], expected_periods)
        np.testing.assert_array_almost_equal(result['powers'], expected_powers, decimal=10)
        
        
        for i in range(6):
            row_data = result['values'][i]
            period = result['periods'][i]
            power = result['powers'][i]
            
            
            single_result = ta_indicators.vpwma(close, period=period, power=power)
            
            assert_close(
                row_data,
                single_result,
                rtol=1e-10,
                msg=f"Batch row {i} (period={period}, power={power}) mismatch"
            )
    
    def test_vpwma_batch_edge_cases(self, test_data):
        """Test VPWMA batch edge cases"""
        close = test_data['close'][:50]
        
        
        result = ta_indicators.vpwma_batch(
            close,
            period_range=(14, 14, 1),
            power_range=(0.382, 0.382, 0.1)
        )
        assert result['values'].shape[0] == 1
        
        
        result = ta_indicators.vpwma_batch(
            close,
            period_range=(10, 12, 5),  
            power_range=(0.3, 0.3, 0)
        )
        
        assert result['values'].shape[0] == 1
        assert result['periods'][0] == 10
        
        
        result = ta_indicators.vpwma_batch(
            close,
            period_range=(14, 14, 0),
            power_range=(0.382, 0.382, 0)
        )
        assert result['values'].shape[0] == 1
        
        
        result = ta_indicators.vpwma_batch(
            close,
            period_range=(10, 14, 2),  
            power_range=(0.382, 0.382, 0)  
        )
        assert result['values'].shape[0] == 3
        assert len(np.unique(result['powers'])) == 1
    
    def test_vpwma_all_nan_input(self):
        """Test VPWMA with all NaN values"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.vpwma(all_nan, period=14, power=0.382)
    
    def test_vpwma_batch_invalid_params(self):
        """Test VPWMA batch with invalid parameters"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0])
        
        
        with pytest.raises(ValueError, match="Not enough valid data|Invalid period"):
            ta_indicators.vpwma_batch(
                data,
                period_range=(20, 25, 5),  
                power_range=(0.382, 0.382, 0)
            )
        
        
        with pytest.raises(ValueError, match="Invalid power"):
            ta_indicators.vpwma_batch(
                data,
                period_range=(5, 5, 0),
                power_range=(float('nan'), float('nan'), 0)
            )
        
        
        with pytest.raises(ValueError, match="Input data slice is empty"):
            ta_indicators.vpwma_batch(
                np.array([]),
                period_range=(14, 14, 0),
                power_range=(0.382, 0.382, 0)
            )
    
    def test_vpwma_batch_metadata(self, test_data):
        """Test VPWMA batch metadata extraction"""
        close = test_data['close'][:100]
        
        result = ta_indicators.vpwma_batch(
            close,
            period_range=(12, 16, 2),    
            power_range=(0.3, 0.5, 0.2)  
        )
        
        
        assert 'values' in result
        assert 'periods' in result
        assert 'powers' in result
        
        
        num_combos = len(result['periods'])
        assert num_combos == 6  
        assert result['values'].shape == (num_combos, len(close))
        
        
        for i in range(num_combos):
            period = result['periods'][i]
            row_data = result['values'][i]
            expected_warmup = period - 1
            
            
            assert np.all(np.isnan(row_data[:expected_warmup])), \
                f"Expected NaN in warmup for combo {i} (period={period})"
            
            
            if len(row_data) > expected_warmup:
                assert not np.isnan(row_data[expected_warmup]), \
                    f"Expected valid value after warmup for combo {i}"
    
    def test_vpwma_kernel_selection(self, test_data):
        """Test VPWMA with different kernel selections"""
        close = test_data['close'][:100]  
        
        
        kernels = ['auto', 'scalar', 'avx2', 'avx512']
        results = {}
        available_kernels = []
        
        for kernel in kernels:
            try:
                results[kernel] = ta_indicators.vpwma(
                    close, 
                    period=14, 
                    power=0.382,
                    kernel=kernel
                )
                available_kernels.append(kernel)
            except ValueError as e:
                
                msg = str(e)
                allowed = (
                    "Unknown kernel" in msg or
                    "not available" in msg.lower() or
                    "not compiled in this build" in msg
                )
                if not allowed:
                    raise
        
        
        assert 'auto' in available_kernels, "Auto kernel should always be available"
        assert 'scalar' in available_kernels, "Scalar kernel should always be available"
        
        
        if len(results) > 1:
            scalar_result = results['scalar']
            for kernel, result in results.items():
                if kernel != 'scalar':
                    
                    valid_mask = ~np.isnan(scalar_result)
                    assert_close(
                        result[valid_mask],
                        scalar_result[valid_mask],
                        rtol=1e-9,  
                        msg=f"Kernel {kernel} mismatch with scalar"
                    )
        
        
        if 'auto' in results:
            assert len(results['auto']) == len(close), "Auto kernel should produce full output"


if __name__ == '__main__':
    pytest.main([__file__, '-v'])

File: tests/wasm/test_vpwma.js
/**
 * WASM binding tests for VPWMA indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/my_project.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('VPWMA partial params', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.vpwma_js(close, 14, 0.382);
    assert.strictEqual(result.length, close.length);
});

test('VPWMA accuracy', async () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.vpwma;
    
    const result = wasm.vpwma_js(
        close,
        expected.defaultParams.period,
        expected.defaultParams.power
    );
    
    assert.strictEqual(result.length, close.length);
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-4,  
        "VPWMA last 5 values mismatch"
    );
    
    
    await compareWithRust('vpwma', result, 'close', expected.defaultParams);
});

test('VPWMA zero period', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.vpwma_js(inputData, 0, 0.382);
    }, /Invalid period/);
});

test('VPWMA period exceeds length', () => {
    
    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.vpwma_js(dataSmall, 10, 0.382);
    }, /Invalid period/);
});

test('VPWMA very small dataset', () => {
    
    const singlePoint = new Float64Array([42.0]);
    
    assert.throws(() => {
        wasm.vpwma_js(singlePoint, 2, 0.382);
    }, /Invalid period|Not enough valid data/);
});

test('VPWMA empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.vpwma_js(empty, 14, 0.382);
    }, /Input data slice is empty/);
});

test('VPWMA invalid power', () => {
    
    const data = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0]);
    
    
    assert.throws(() => {
        wasm.vpwma_js(data, 2, NaN);
    }, /Invalid power/);
    
    
    assert.throws(() => {
        wasm.vpwma_js(data, 2, Infinity);
    }, /Invalid power/);
});

test('VPWMA reinput', () => {
    
    const close = new Float64Array(testData.close);
    
    
    const firstResult = wasm.vpwma_js(close, 14, 0.382);
    assert.strictEqual(firstResult.length, close.length);
    
    
    const secondResult = wasm.vpwma_js(firstResult, 5, 0.5);
    assert.strictEqual(secondResult.length, firstResult.length);
    
    
    if (secondResult.length > 240) {
        for (let i = 240; i < secondResult.length; i++) {
            assert(!isNaN(secondResult[i]), `Unexpected NaN at index ${i}`);
        }
    }
});

test('VPWMA NaN handling', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.vpwma_js(close, 14, 0.382);
    assert.strictEqual(result.length, close.length);
    
    
    if (result.length > 50) {
        for (let i = 50; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }
    
    
    assertAllNaN(result.slice(0, 13), "Expected NaN in warmup period");
});

test('VPWMA all NaN input', () => {
    
    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);
    
    assert.throws(() => {
        wasm.vpwma_js(allNaN, 14, 0.382);
    }, /All values are NaN/);
});

test('VPWMA batch single parameter set', () => {
    
    const close = new Float64Array(testData.close);
    
    
    const batchResult = wasm.vpwma_batch(close, {
        period_range: [14, 14, 0],
        power_range: [0.382, 0.382, 0]
    });
    
    
    const singleResult = wasm.vpwma_js(close, 14, 0.382);
    
    assert.strictEqual(batchResult.values.length, singleResult.length);
    assertArrayClose(batchResult.values, singleResult, 1e-10, "Batch vs single mismatch");
});

test('VPWMA batch multiple periods', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100)); 
    
    
    const batchResult = wasm.vpwma_batch(close, {
        period_range: [14, 18, 2],      
        power_range: [0.382, 0.382, 0]  
    });
    
    
    assert.strictEqual(batchResult.values.length, 3 * 100);
    assert.strictEqual(batchResult.rows, 3);
    assert.strictEqual(batchResult.cols, 100);
    
    
    const periods = [14, 16, 18];
    for (let i = 0; i < periods.length; i++) {
        const rowStart = i * 100;
        const rowEnd = rowStart + 100;
        const rowData = batchResult.values.slice(rowStart, rowEnd);
        
        const singleResult = wasm.vpwma_js(close, periods[i], 0.382);
        assertArrayClose(
            rowData, 
            singleResult, 
            1e-10, 
            `Period ${periods[i]} mismatch`
        );
    }
});

test('VPWMA batch metadata from result', () => {
    
    const close = new Float64Array(20); 
    close.fill(100);
    
    const result = wasm.vpwma_batch(close, {
        period_range: [14, 18, 2],      
        power_range: [0.3, 0.5, 0.1]   
    });
    
    
    assert.strictEqual(result.combos.length, 9);
    
    
    assert.strictEqual(result.combos[0].period, 14);   
    assert.strictEqual(result.combos[0].power, 0.3);  
    
    
    assert.strictEqual(result.combos[8].period, 18);  
    assertClose(result.combos[8].power, 0.5, 1e-10, "power mismatch"); 
});

test('VPWMA batch full parameter sweep', () => {
    
    const close = new Float64Array(testData.close.slice(0, 50));
    
    const batchResult = wasm.vpwma_batch(close, {
        period_range: [12, 14, 2],      
        power_range: [0.3, 0.4, 0.1]    
    });
    
    
    assert.strictEqual(batchResult.combos.length, 4);
    assert.strictEqual(batchResult.rows, 4);
    assert.strictEqual(batchResult.cols, 50);
    assert.strictEqual(batchResult.values.length, 4 * 50);
    
    
    for (let combo = 0; combo < batchResult.combos.length; combo++) {
        const period = batchResult.combos[combo].period;
        const power = batchResult.combos[combo].power;
        
        const rowStart = combo * 50;
        const rowData = batchResult.values.slice(rowStart, rowStart + 50);
        
        
        for (let i = 0; i < period - 1; i++) {
            assert(isNaN(rowData[i]), `Expected NaN at warmup index ${i} for period ${period}`);
        }
        
        
        for (let i = period - 1; i < 50; i++) {
            assert(!isNaN(rowData[i]), `Unexpected NaN at index ${i} for period ${period}`);
        }
    }
});

test('VPWMA batch edge cases', () => {
    
    const close = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    
    
    const singleBatch = wasm.vpwma_batch(close, {
        period_range: [5, 5, 1],
        power_range: [0.382, 0.382, 0.1]
    });
    
    assert.strictEqual(singleBatch.values.length, 10);
    assert.strictEqual(singleBatch.combos.length, 1);
    
    
    const largeBatch = wasm.vpwma_batch(close, {
        period_range: [5, 7, 10], 
        power_range: [0.382, 0.382, 0]
    });
    
    
    assert.strictEqual(largeBatch.values.length, 10);
    assert.strictEqual(largeBatch.combos.length, 1);
    
    
    assert.throws(() => {
        wasm.vpwma_batch(new Float64Array([]), {
            period_range: [14, 14, 0],
            power_range: [0.382, 0.382, 0]
        });
    }, /Input data slice is empty/);
});


test('VPWMA batch - new ergonomic API with single parameter', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.vpwma_batch(close, {
        period_range: [14, 14, 0],
        power_range: [0.382, 0.382, 0]
    });
    
    
    assert(result.values, 'Should have values array');
    assert(result.combos, 'Should have combos array');
    assert(typeof result.rows === 'number', 'Should have rows count');
    assert(typeof result.cols === 'number', 'Should have cols count');
    
    
    assert.strictEqual(result.rows, 1);
    assert.strictEqual(result.cols, close.length);
    assert.strictEqual(result.combos.length, 1);
    assert.strictEqual(result.values.length, close.length);
    
    
    const combo = result.combos[0];
    assert.strictEqual(combo.period, 14);
    assert.strictEqual(combo.power, 0.382);
    
    
    const oldResult = wasm.vpwma_js(close, 14, 0.382);
    for (let i = 0; i < oldResult.length; i++) {
        if (isNaN(oldResult[i]) && isNaN(result.values[i])) {
            continue; 
        }
        assert(Math.abs(oldResult[i] - result.values[i]) < 1e-10,
               `Value mismatch at index ${i}`);
    }
});

test('VPWMA batch - new API with multiple parameters', () => {
    
    const close = new Float64Array(testData.close.slice(0, 50));
    
    const result = wasm.vpwma_batch(close, {
        period_range: [10, 14, 2],       
        power_range: [0.3, 0.4, 0.1]     
    });
    
    
    assert.strictEqual(result.combos.length, 6);
    assert.strictEqual(result.rows, 6);
    assert.strictEqual(result.cols, 50);
    
    
    const expectedCombos = [
        {period: 10, power: 0.3},
        {period: 10, power: 0.4},
        {period: 12, power: 0.3},
        {period: 12, power: 0.4},
        {period: 14, power: 0.3},
        {period: 14, power: 0.4}
    ];
    
    for (let i = 0; i < expectedCombos.length; i++) {
        assert.strictEqual(result.combos[i].period, expectedCombos[i].period);
        assertClose(result.combos[i].power, expectedCombos[i].power, 1e-10, `Power mismatch at combo ${i}`);
    }
});

test.skip('VPWMA zero-copy API', () => {
    
    
    
    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]);
    const period = 5;
    const power = 0.382;
    
    
    const ptr = wasm.vpwma_alloc(data.length);
    assert(ptr !== 0, 'Should allocate memory successfully');
    
    try {
        
        const memory = wasm.memory;
        const memView = new Float64Array(memory.buffer, ptr, data.length);
        memView.set(data);
        
        
        wasm.vpwma_into(ptr, ptr, data.length, period, power);
        
        
        const regularResult = wasm.vpwma_js(data, period, power);
        
        for (let i = 0; i < data.length; i++) {
            if (isNaN(regularResult[i]) && isNaN(memView[i])) {
                continue; 
            }
            assert(Math.abs(regularResult[i] - memView[i]) < 1e-10,
                   `Zero-copy mismatch at index ${i}: regular=${regularResult[i]}, zerocopy=${memView[i]}`);
        }
    } finally {
        wasm.vpwma_free(ptr, data.length);
    }
});

test.skip('VPWMA zero-copy with large dataset', () => {
    
    
    const size = 10000;
    const data = new Float64Array(size);
    for (let i = 0; i < size; i++) {
        data[i] = Math.sin(i * 0.1) * 100 + 100;
    }
    
    const ptr = wasm.vpwma_alloc(size);
    assert(ptr !== 0, 'Should allocate large memory successfully');
    
    try {
        const memory = wasm.memory;
        const memView = new Float64Array(memory.buffer, ptr, size);
        memView.set(data);
        
        
        wasm.vpwma_into(ptr, ptr, size, 14, 0.382);
        
        
        for (let i = 0; i < 13; i++) {
            assert(isNaN(memView[i]), `Expected NaN at warmup index ${i}`);
        }
        
        
        for (let i = 13; i < 100; i++) {
            assert(!isNaN(memView[i]), `Unexpected NaN at index ${i}`);
        }
    } finally {
        wasm.vpwma_free(ptr, size);
    }
});


test.skip('VPWMA zero-copy error handling', () => {
    
    
    assert.throws(() => {
        wasm.vpwma_into(0, 0, 10, 14, 0.382);
    }, /null pointer|invalid memory/i);
    
    const ptr = wasm.vpwma_alloc(10);
    try {
        
        assert.throws(() => {
            wasm.vpwma_into(ptr, ptr, 10, 0, 0.382);
        }, /Invalid period/);
        
        
        assert.throws(() => {
            wasm.vpwma_into(ptr, ptr, 10, 5, NaN);
        }, /Invalid power/);
    } finally {
        wasm.vpwma_free(ptr, 10);
    }
});

test.skip('VPWMA zero-copy memory management', () => {
    
    
    const sizes = [10, 100, 1000];
    
    for (const size of sizes) {
        const ptr = wasm.vpwma_alloc(size);
        assert(ptr !== 0, `Should allocate ${size} elements`);
        
        
        const memory = wasm.memory;
        const memView = new Float64Array(memory.buffer, ptr, size);
        for (let i = 0; i < size; i++) {
            memView[i] = i;
        }
        
        
        for (let i = 0; i < size; i++) {
            assert.strictEqual(memView[i], i, `Memory corruption at index ${i}`);
        }
        
        
        wasm.vpwma_free(ptr, size);
    }
});

test('VPWMA warmup period verification', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100));
    
    const testCases = [
        {period: 5, power: 0.5, expectedWarmup: 4},
        {period: 10, power: 0.382, expectedWarmup: 9},
        {period: 14, power: 0.382, expectedWarmup: 13},
        {period: 20, power: 0.3, expectedWarmup: 19}
    ];
    
    for (const tc of testCases) {
        const result = wasm.vpwma_js(close, tc.period, tc.power);
        
        
        for (let i = 0; i < tc.expectedWarmup; i++) {
            assert(isNaN(result[i]), 
                   `Expected NaN at index ${i} for period=${tc.period}, got ${result[i]}`);
        }
        
        
        assert(!isNaN(result[tc.expectedWarmup]), 
               `Expected first valid value at index ${tc.expectedWarmup} for period=${tc.period}`);
        
        
        assert.strictEqual(tc.expectedWarmup, tc.period - 1,
                          `Warmup calculation mismatch for period=${tc.period}`);
    }
});

test('VPWMA SIMD consistency', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100));
    const period = 14;
    const power = 0.382;
    
    
    const kernels = ['scalar', 'auto'];
    const results = {};
    
    for (const kernel of kernels) {
        try {
            
            
            
            results[kernel] = wasm.vpwma_js(close, period, power);
        } catch (e) {
            
            continue;
        }
    }
    
    
    assert(results['auto'], 'Auto kernel should be available');
    
    
    if (Object.keys(results).length > 1) {
        const baseResult = results['auto'];
        for (const [kernel, result] of Object.entries(results)) {
            if (kernel === 'auto') continue;
            
            for (let i = 0; i < result.length; i++) {
                if (isNaN(baseResult[i]) && isNaN(result[i])) {
                    continue;
                }
                assert(Math.abs(baseResult[i] - result[i]) < 1e-9,
                       `Kernel ${kernel} mismatch with auto at index ${i}`);
            }
        }
    }
});



test.after(() => {
    console.log('VPWMA WASM tests completed');
});